### PR TITLE
Limit data read during operation

### PR DESCRIPTION
**Pull request by AlCutter**
Spanner seems to not take an error being returned by a RowIterator func as a reason to cancel/abandon the underlying read, and the encapsulating transaction will not commit until the read has finished and so this causes poor performance, particularly during sequencing.

This change uses `ReadOptions.Limit` to specify the number of rows to read when dequeueing leaves, and switches to SQL for fetching subtree data.

#950

### CODE REVIEW

File: storage/cloudspanner/log_storage.go [hunk 1]
```diff
@@ -536,11 +536,14 @@ func (tx *logTX) DequeueLeaves(ctx context.Context, limit int, cutoff time.Time)
 		rows = genDequeueRowsOld(tx.getLogStorageConfig(), tx.treeID, now.Unix(), rand.Int31n(numByteValues))
 	}
 	ret := make([]*trillian.LogLeaf, 0, limit)
+	readOpts := &spanner.ReadOptions{Limit: limit}
+	drain := 0
```
Comment (Martin2112) [inline storage/cloudspanner/log_storage.go]: 'discarded' maybe? Wasn't immediately clear what it meant.

File: storage/cloudspanner/log_storage.go [hunk 2]
```diff
@@ -572,6 +578,9 @@ func (tx *logTX) DequeueLeaves(ctx context.Context, limit int, cutoff time.Time)
 			break
 		}
 	}
+	if drain > 0 {
+		glog.Warningf("Drained %d", drain)
```
Comment (Martin2112) [inline storage/cloudspanner/log_storage.go]: Is this bad? Can you expand on the message if necessary.

File: storage/cloudspanner/spanner.sdl [hunk 3]
```diff
@@ -29,10 +29,7 @@ CREATE TABLE TreeHeads(
   RootSignature           BYTES(1024) NOT NULL,
   TreeRevision            INT64 NOT NULL,
   TreeMetadata            BYTES(2097152),
-) PRIMARY KEY(TreeID, TimestampNanos DESC);
-
-CREATE UNIQUE INDEX TreeRevisionIndex
-  ON TreeHeads(TreeID, TreeRevision DESC);
+) PRIMARY KEY(TreeID, TreeRevision DESC);
```
Comment (Martin2112) [inline storage/cloudspanner/spanner.sdl]: Just for the release notes this looks like an incompatible change?

File: storage/cloudspanner/tree_storage.go [hunk 4]
```diff
@@ -391,12 +408,10 @@ func (t *treeTX) getSubtree(ctx context.Context, rev int64, id storage.NodeID) (
...
 		}
-		// We've got what we want, tell spanner to stop reading by returning
-		// not-really-an-error:
-		return errFinished
+		return nil
 	})
-	if err == errFinished {
-		err = nil
+	if drain > 0 {
+		glog.Warningf("getSubtree drained %d", drain)
```
Comment (Martin2112) [inline storage/cloudspanner/tree_storage.go]: Not sure how you can get this given the `LIMIT 1` per subtree?

File: storage/cloudspanner/log_storage.go (hunk 1 above)
Comment (AlCutter) [reply storage/cloudspanner/log_storage.go]: I'll remove it - I was just verifying that it did what I expected.

File: storage/cloudspanner/log_storage.go (hunk 2 above)
Comment (AlCutter) [reply storage/cloudspanner/log_storage.go]: removed

File: storage/cloudspanner/spanner.sdl (hunk 3 above)
Comment (AlCutter) [reply storage/cloudspanner/spanner.sdl]: Technically yes, but presumably nobody other than me has an instance of it :)

File: storage/cloudspanner/tree_storage.go (hunk 4 above)
Comment (AlCutter) [reply storage/cloudspanner/tree_storage.go]: removed - I was just verifying that.

**Comment by AlCutter** [comment]
PTAL

File: storage/cloudspanner/spanner.sdl (hunk 3 above)
Comment (Martin2112) [reply storage/cloudspanner/spanner.sdl]: Added to draft release notes in case.

**Pull request merged**